R/do_findChromPeaks-functions.R

```diff
@@ -2220,9 +2220,220 @@ do_findPeaks_MSW <- function(mz, int, snthresh = 3,
     peaklist
 }
 
+############################################################
+## Ion-mobility peak picking
+##
+#' @title Core API for Centwave-based ion-mobility peak picking 
+#' @name do_findChromPeaks_IM_centWave
+#' 
+#' @description Performs an extension of CentWave peak-picking on LC-IM-MS MS1
+#'   data. First it joins all scans into frames and performs .centWave_orig on
+#'   the summarized LC-MS-like data. From each peak, it calculates its mobilogram and
+#'   performs a second peak-picking on the IM dimension, resolving the peaks.
+#' 
+#' @inheritParams do_findChromPeaks_centWave
+#' @inheritParams findChromPeaks-centWaveIonMobility
+#' 
+#' @return A matrix, each row representing an identified peak, with columns:
+#'     \describe{
+#'     \item{mz}{m/z value of the peak at the apex position.}
+#'     \item{mzmin}{Minimum m/z of the peak.}
+#'     \item{mzmax}{Maximum m/z of the peak.}
+#'     \item{rt}{Retention time value of the peak at the apex position.}
+#'     \item{rtmin}{Minimum retention time of the peak.}
+#'     \item{rtmax}{Maximum retention time of the peak.}
+#'     \item{im}{Ion mobility value of the peak at the apex position.}
+#'     \item{immin}{Minimum ion mobility value of the peak.}
+#'     \item{immax}{Maximum ion mobility value of the peak.}
+#'     \item{maxo}{Maximum intensity of the peak.}
+#'     \item{into}{Integrated (original) intensity of the peak.}
+#'     \item{intb}{Always \code{NA}.}
+#'     \item{sn}{Always \code{NA}}
+#'     }
+#'
+#' @family core peak detection functions
+#' 
+#' @author Roger Gine, Johannes Rainer
+#' 
+#' @importFrom Spectra peaksData rtime combineSpectra mz
+do_findChromPeaks_IM_centWave <- function(spec,
```

Comment:
Are there other ways/algorithms to perform peak detection on IM data that do not first collapse the data and then expand it again like you do here?

If not I would suggest to split the functionality into 3 functions:

1) collapse `peaksData` by frame
2) do peak detection using the original detection algorithm.
3) have a function that takes the original `peaksData` and the detected chrom peaks matrix from 2) as input and *post processes* the data.

Reply:
In principle, yes, there are other ways to perform peak-picking that directly use the "full" data without collapsing (for instance, you could do a 2D-CWT where you change the scales in both RT and IM dimensions and find local maxima; or any peakpicking algorithm such as those used for GCxGC-MS, where you have a similar situation). I haven't looked deeply into such methods, but we should accomodate for them too, just in case

Still, splitting the functionality in `do_findChromPeaks_IM_centWave` seems good, since those functions would be reusable for other algorithms, etc.  Specifically, if you agree, I'll do the following: 
- Call `.mse_find_chrom_peaks_sample` with the IM-collapsed `Spectra` object and the param `as(param, CentWaveParam")` (since it `IMCentWaveParam` inherits from it) -> That function will take care of extracting the peaksData, rt, valsPerSpect, etc., run `do_findChromPeaks_centWave`,  and return the peak matrix.
- Encapsulate the post processing (ie. resolving across IM dimension) in another function

Sounds good?  :+1:

Reply:
sorry for my late reply!

splitting functionality is always good, as you say, enables reuse - and makes the code easier to read. so, yes, it sounds good.

so, if I understand: 

- if `param` inherits `IMCentWaveParam` collapse the `peaksData` by frame.
- pass the data to `do_findChromPeaks_centWave` for peak detection (since `IMCentWaveParam` inherits `CentWaveParam`)
- if `param` inherits `IMCentWaveParam` post process the detected peaks (resolve across IM dimension) and return results.

if the functions get to large, you could also consider implementing a `.im_mse_find_chrom_peaks_sample` that is called instead of `.mse_find_chrom_peaks_sample` if `param` inherits from an `IM` param object... not sure if that would simplify integration of additional/other IM peak detection methods.

Reply:
For now I've encapsulated all this into `do_findChromPeaks_IM_centWave`, so it's called _after_ dispatching the function corresponding to the `param`:

``` 
    ## Merging all frames from the same scans to summarize across IM dimension
    scans_summarized <-
        Spectra::combineSpectra(
            spec,
            f = as.factor(spec$frameId),
            intensityFun = base::sum,
            weighted = TRUE,
            ppm = ppmMerging
        )
    Spectra::centroided(scans_summarized) <- TRUE
    
    ## 1D Peak-picking on summarized data
    peaks <- .mse_find_chrom_peaks_sample(scans_summarized,
                                          msLevel = 1L,
                                          param = CentWaveParam(ppm = ppm, peakwidth = peakwidth,
                                                        snthresh = snthresh, prefilter = prefilter,
                                                        mzCenterFun = mzCenterFun, integrate = integrate,
                                                        mzdiff = mzdiff, fitgauss = fitgauss, noise = noise,
                                                        verboseColumns = verboseColumns, roiList = roiList,
                                                        firstBaselineCheck = firstBaselineCheck,
                                                        roiScales = roiScales,
                                                        extendLengthMSW = extendLengthMSW))
    
    ## 1D Peak-picking, for each individual peak, to resolve across the IM dimension
    .do_resolve_IM_peaks_CWT(spec, peaks, binWidthIM)
```

It's basically the all steps you mentioned (collapse, CentWave and resolve), but called from a "lower" function call level, so everything upstream is more tidy. What I like about this is that all Centwave-specific checks and data extraction (mz, int, valsPerSpect, etc.) are handled by `.mse_find_chrom_peaks_sample`, so we are reusing already-existing code.

I'll commit the proposed refactor so you can take a closer look by yourself

Reply:
Excellent. Let me know when you're OK/ready from your side. I would then like to try the code in action and tinker myself a bit to see if/where we could improve/optimize.

For that I would create yet another branch to play with the code and ask for your feedback on the merge.

Related to that: could you please provide a short code snipped with the example how to perform the analysis (I guess I got already a file from you on which I can test...)?

Reply:
Definitely! If you agree, perhaps we should use (for testing) the `MsBackendMemory` or `MsBackendDataFrame` to create a manageable toy example from the current _read-only_ `MsBackendTimsTof` (for instance, subsetting the RT)

I'm still figuring out how to properly centroid the data (the format has some problems that makes the current `peakPicks` function in `Spectra` ineffective, see below), but still, the analysis would go somewhat like this:

``` r 
library(opentimsr)
library(MsBackendTimsTof)
library(MsExperiment)
library(Spectra)
library(xcms)
library(magrittr)

## Setting up Bruker SDK to read the raw data directly (if you have it already, just point to the corresponding file)
so_folder <- tempdir()
so_file <- download_bruker_proprietary_code(so_folder, method = "wget")
setup_bruker_so(so_file)

## Set up valid MsExperiment with the subsetted spectra, then detect IM peaks
be <- backendInitialize(MsBackendTimsTof(), "./path_to_your_file_folder.d")
spec <- Spectra(be) %>%
    filterRt(., c(350, 365)) %>% 
    filterMsLevel(.. 1) %>% 
    setBackend(., MsBackendMemory())

exp <- MsExperiment()
spectra(exp) <- spec
sampleData(exp) <- DataFrame(
    raw_file = normalizePath("./path_to_your_file_folder.d")
)
exp <- linkSampleData(exp, with = "sampleData.raw_file = spectra.dataOrigin")

exp <- findChromPeaks(exp, IMCentWaveParam())
```

You can use the TIMS-TOF data file I sent you a while back, just bear in mind it's in a zero-less profile mode (working on fixing that) and the chromatographic peaks are usually very short (<6-10s)